tests/test_file_watcher.py

```diff
@@ -0,0 +1,181 @@
+    def tearDownClass(cls) -> None:
+        # Don't call the superclass.
+        pass
+
+    def setUp(self) -> Generator:
+        self.assertEqual(len(TestFileWatcher._active_watchers), 0)
+        # Watchers are only registered when there are workspace folders so add a folder.
+        self.folder_root_path = setup_workspace_folder()
+        yield from super().setUpClass()
+        yield from super().setUp()
+
+    def tearDown(self) -> Generator:
+        yield from super().tearDownClass()
+        self.assertEqual(len(TestFileWatcher._active_watchers), 0)
+        # Restore original project data.
+        window = sublime.active_window()
+        window.set_project_data(None)
+
+
+class FileWatcherStaticTests(FileWatcherDocumentTestCase):
+
+    @classmethod
+    def get_stdio_test_config(cls) -> ClientConfig:
+        return ClientConfig.from_config(
+            super().get_stdio_test_config(),
+            {
+                'file_watcher': {
+                    'glob': '*.js',
+                    'kind': ['change'],
+                    'ignores': ['.git'],
+                }
+            }
+        )
+
+    def test_initialize_params_includes_capability(self) -> None:
+        self.assertIn('didChangeWatchedFiles', self.initialize_params['capabilities']['textDocument'])
+
+    def test_creates_a_static_watcher(self) -> None:
+        # Starting a session should have created a watcher.
+        self.assertEqual(len(TestFileWatcher._active_watchers), 1)
+        watcher = TestFileWatcher._active_watchers[0]
+        self.assertEqual(watcher.glob, '*.js')
+        self.assertEqual(watcher.kind, ['change'])
+        self.assertEqual(watcher.ignores, ['.git'])
+        self.assertEqual(watcher.root_path, self.folder_root_path)
+
+    def test_handles_file_event(self) -> Generator:
+        watcher = TestFileWatcher._active_watchers[0]
+        filepath = join(self.folder_root_path, 'file.js')
+        watcher.trigger_event([('change', filepath)])
+        sent_notification = yield from self.await_message('workspace/didChangeWatchedFiles')
+        self.assertIs(type(sent_notification['changes']), list)
+        self.assertEqual(len(sent_notification['changes']), 1)
+        change = sent_notification['changes'][0]
+        self.assertEqual(change['type'], file_watcher_kind_to_lsp_file_change_type('change'))
+        self.assertTrue(change['uri'].endswith('file.js'))
+
+
+class FileWatcherDynamicTests(FileWatcherDocumentTestCase):
+
+    def test_handles_dynamic_watcher_registration(self) -> Generator:
+        registration_params = {
+            'registrations': [
+                {
+                    'id': '111',
+                    'method': 'workspace/didChangeWatchedFiles',
+                    'registerOptions': {
+                        'watchers': [
+                            {
+                                'globPattern': '*.py',
+                                'kind': WatchKindCreate | WatchKindChange,
```

Comment:
Maybe use 
```
FileChangeTypeCreated = 1
FileChangeTypeChanged = 2
```
instead of 
```
WatchKindCreate = 1
WatchKindChange = 2
```
?

Reply:
For the concrete test, it would make no difference what ever you choose.

But if someone would add `WatchKindDelete` than it would cause a problem. 
But because this is a test, I guess it is not that important.

Reply:
I will add `WatchKindDelete` to the test also.

But I can't use `FileChangeType` in place of `WatchKind` as those have different purpose. The latter is used in a bit field when registering the watcher dynamically. The former is a single value used to notify about file event through the `workspace/didChangeWatchedFiles` notification.